src/modules/ekf2/EKF/aid_sources/range_finder/range_finder_consistency_check.cpp

```diff
@@ -36,56 +36,107 @@
  */
 
 #include <aid_sources/range_finder/range_finder_consistency_check.hpp>
+#include "ekf_derivation/generated/range_validation_filter.h"
 
-void RangeFinderConsistencyCheck::update(float dist_bottom, float dist_bottom_var, float vz, float vz_var,
-		bool horizontal_motion, uint64_t time_us)
+using namespace matrix;
+
+void RangeFinderConsistencyCheck::init(const float z, const float z_var, const float dist_bottom,
+				       const float dist_bottom_var)
 {
-	if (horizontal_motion) {
-		_time_last_horizontal_motion = time_us;
-	}
+	_R.setZero();
+	_A.setIdentity();
+	float p[4] = {z_var, 0.f, 0.f, z_var + dist_bottom_var};
+	_P = Matrix<float, RangeFilter::size, RangeFilter::size>(p);
+	sym::RangeValidationFilter(&_H);
+	_x(RangeFilter::z.idx) = z;
+	_x(RangeFilter::terrain.idx) = z - dist_bottom;
+	_initialized = true;
+	_sample_count = 0;
+	_state = _test_ratio_lpf.getState() < 1.f ? KinematicState::UNKNOWN : KinematicState::INCONSISTENT;
+	_test_ratio_lpf.reset(2.f);
+}
 
+void RangeFinderConsistencyCheck::update(const float z, const float z_var, const float vz, const float vz_var,
+		const float dist_bottom, const float dist_bottom_var, const uint64_t time_us)
+{
 	const float dt = static_cast<float>(time_us - _time_last_update_us) * 1e-6f;
 
-	if ((_time_last_update_us == 0)
-	    || (dt < 0.001f) || (dt > 0.5f)) {
+	if (_time_last_update_us == 0 || dt > 1.f) {
 		_time_last_update_us = time_us;
-		_dist_bottom_prev = dist_bottom;
+		init(z, z_var, dist_bottom, dist_bottom_var);
 		return;
 	}
 
-	const float vel_bottom = (dist_bottom - _dist_bottom_prev) / dt;
-	_innov = -vel_bottom - vz; // vel_bottom is +up while vz is +down
+	if (_min_nr_of_samples == 0) {
+		_min_nr_of_samples = (int)(1.f / dt);
+	}
+
+	_time_last_update_us = time_us;
 
-	// Variance of the time derivative of a random variable: var(dz/dt) = 2*var(z) / dt^2
-	const float var = 2.f * dist_bottom_var / (dt * dt);
-	_innov_var = var + vz_var;
+	_R(RangeFilter::z.idx, RangeFilter::z.idx) = z_var;
+	_R(RangeFilter::terrain.idx, RangeFilter::terrain.idx) = dist_bottom_var;
 
-	const float normalized_innov_sq = (_innov * _innov) / _innov_var;
-	_test_ratio = normalized_innov_sq / (_gate * _gate);
-	_signed_test_ratio_lpf.setParameters(dt, _signed_test_ratio_tau);
-	const float signed_test_ratio = matrix::sign(_innov) * _test_ratio;
-	_signed_test_ratio_lpf.update(signed_test_ratio);
+	SquareMatrix<float, 2> Q;
+	Q(RangeFilter::z.idx, RangeFilter::z.idx) = dt * dt * vz_var + 0.001f;
+	Q(RangeFilter::terrain.idx, RangeFilter::terrain.idx) = terrain_process_noise;
 
-	updateConsistency(vz, time_us);
+	_x(RangeFilter::z.idx) += dt * vz;
+	_P = _A * _P * _A.transpose() + Q;
```

Comment:
`_A` is just the identity matrix, right? So you could just do `_P += Q` and not store _A as a member variable

Reply:
Or even directly apply the noise to P the diagonal elements of P instead of first setting them in Q fillowed by a matrix addition. e.g.: `_P(RangeFilter::terrain.idx, RangeFilter::terrain.idx) = terrain_process_noise;`